feat: add `wallet_requestExecutionPermissions` method

jeffsmale90 opened the pull request:
Adds the method `wallet_requestExecutionPermissions` which is defined in this proposed revision of the EIP-7715 specification https://github.com/ethereum/ERCs/pull/1098.

## **Description**

This method is proxied to `@metamask/permissions-kernel-snap`, which implements the exact same method. This snap will be preinstalled in Extension, before the feature is enabled for any users.

The feature is gated by `process.env.EIP_7715_READABLE_PERMISSIONS_ENABLED` and requires `process.env.PERMISSIONS_KERNEL_SNAP_ID` to be set. Presently `EIP_7715_READABLE_PERMISSIONS_ENABLED` is set to false, and `PERMISSIONS_KERNEL_SNAP_ID` is set to an empty string for all builds.

Requires https://github.com/MetaMask/eth-json-rpc-middleware/pull/396 released in https://github.com/MetaMask/eth-json-rpc-middleware/releases/tag/v17.1.0. This change also adds support for `wallet_revokePermission`, which will be implemented in a future PR, once the revocation UI is implemented.


[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/35193?quickstart=1)

## **Changelog**

CHANGELOG entry: As this is behind a local feature gate, there are no public facing changes.

## **Manual testing steps**

Can be tested with https://github.com/MetaMask/delegation-toolkit/pull/60 which adds support to the delegation-toolkit's experimental api to call the wallet method directly, rather than via `wallet_invokeSnap`.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

jeffsmale90 committed:
Add wallet_requestExecutionPermissions method. Pending @metamask/eth-json-rpc-middleware upgrade.

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -2183,6 +2187,15 @@ export default class MetamaskController extends EventEmitter {
         },
         this.controllerMessenger,
       ),
+      processRequestExecutionPermissions:
```
after this feature is enabled and shipped, I'd love to rename this hook to plain old `handleSnapRequest` or something like that

jeffsmale90 replied:
Great idea! I have updated this to `forwardRequestToSnap` which I think is pretty clear about what it's doing. Presently it forwards the request to the specified snap verbatim, but in the future we could support a `mapRequest` function that could accept the incoming request and return the snap request, so that the publicly exposed method doesn't necessarily need to match exactly the snap method.

https://github.com/MetaMask/metamask-extension/pull/35193/files#diff-6fbff2cfe97ac01b77296ef2122c7e0a5b3ff6a84b584b4d1a87482f35eea3d6R2192

lmkwyt

V00D00-child commented on app/scripts/lib/transaction/eip7715.ts:
```diff
@@ -0,0 +1,46 @@
+    );
+  }
+
+  const requestExecutionPermissionsResult = (await handleRequest({
```
Forwarding requests to the `kernel-snap` is a pattern that is similar to how we forward requests to `gator-snap` to fetch granted permissions. Should we move this to the [GatorPermissionsController](https://github.com/MetaMask/core/blob/main/packages/gator-permissions-controller/src/GatorPermissionsController.ts#L283) and expose via an action to the consumer? 

Is there a strong reason why `processRequestExecutionPermissions()` can not live in the `GatorPermissionsController`?

jeffsmale90 replied:
Nice - in response to Jiexi's suggestion above, I've removed the bulk of EIP-7715 bespoke logic, and implemented a generic `forwardRequestToSnap` handler.

The only thing left is Kernel Snap Id, which is provided via environment - I think the Gator Permissions Controller should be accepting the snap Id from the same place as this makes switching between environments really easy.

jeffsmale90 commented:
I have read the CLA Document and I hereby sign the CLA

jeffsmale90 committed:
Update yarn.lock with updated @metamask/eth-json-rpc-middleware

jeffsmale90 committed:
Remove EIP-7715 specific handler, replace with general purpose 'forwardRequestToSnap' handler.

jeffsmale90 committed:
Merge branch 'main' into feat/eip7715_requestExecutionPermissions

jeffsmale90 committed:
Minor fixes:
- Add missing forwardRequestToSnap import to metamask-controller
- Fix linting errors in forwardRequestToSnap and related tests

jeffsmale90 committed:
Minor fixes:
- Add missing forwardRequestToSnap import to metamask-controller
- Fix linting errors in forwardRequestToSnap and related tests
- Deduplicate @metamask/eth-json-rpc-middleware
- Remove requirement for kernel snap Id to be set when checking isGatorPermissionsFeatureEnabled

jeffsmale90 committed:
Merge branch 'main' into feat/eip7715_requestExecutionPermissions

jeffsmale90 committed:
Throw InternalError instead of JsonRpcError, pass Metamask Controller's handleSnapRequest function, rather than binding the call function on the ControllerMessenger

jeffsmale90 committed:
Update forwardRequestToSnap assertions to match thrown errors

jeffsmale90 committed:
Remove errant import from test

jeffsmale90 committed:
Merge branch 'main' into feat/eip7715_requestExecutionPermissions
